server/spec/services/scheduler/strategy/daemon_spec.rb

```diff
@@ -24,6 +20,100 @@
     GridService.create!(name: 'test', grid: grid, image_name: 'foo/bar:latest', stateful: false)
   end
 
+  describe '#find_node' do
+    it 'finds matching nodes' do
+      nodes = scheduler_nodes
+      nodes.each do |n|
+        node = subject.find_node(stateless_service, n.node_number, nodes)
+        node.schedule_counter += 1
+        expect(node).to eq(nodes[n.node_number - 1])
+      end
+    end
+
+    context 'with existing nodes' do
+      before(:each) do
+        host_nodes.each do |n|
+          stateless_service.grid_service_instances.create!(
+            host_node: n,
+            instance_number: n.node_number
+          )
+        end
+      end
+
+      it 'minimizes shuffle when nodes 2 & 3 are missing' do
+        nodes = scheduler_nodes.tap do |n|
+          n.delete_at(1)
+          n.delete_at(1)
+        end
+
+        scheduled = []
+        nodes.each_with_index do |n, i|
+          node = subject.find_node(stateless_service, i + 1, nodes)
+          node.schedule_counter += 1
+          scheduled << node
+        end
+        expect(scheduled.map {|s| s.node_number}).to eq([
+          1, 10, 9, 4, 5, 6, 7, 8
+        ])
+      end
+
+      it 'minimizes shuffle when nodes 2 & 3 are missing (2 instances per node)' do
+        nodes = scheduler_nodes.tap do |n|
+          n.delete_at(1)
+          n.delete_at(1)
+        end
+
+        scheduled = []
+        16.times do |i| # 2 instances per node
+          node = subject.find_node(stateless_service, i + 1, nodes)
+          node.schedule_counter += 1
+          scheduled << node
+        end
+        expect(scheduled.map {|s| s.node_number}).to eq([
+          1, 10, 9, 4, 5, 6, 7, 8,
+          9, 10, 1, 4, 5, 6, 7, 8
```

Comment:
Hmm... strictly speaking shouldn't this case be like:

```ruby
        [
          1, 9, 10, 4, 5, 6, 7, 8, 9, 10,
          1, 7, 8, 4, 5, 6
        ]
```

Now it's still shuffling three instances unnecessarily:
* 14 from node 4 -> 6
* 15 from node 5 -> 7
* 16 from node 6 -> 8

Reply:
Oh... the "2 instances per node" spec only has 10 of the `GridServiceInstances`, not 20?